blog/_src/posts/2016-07-25-tutorial-zero-to-sixty-in-racket.scrbl

```diff
@@ -0,0 +1,698 @@
+      (display (first words))
+      (for ([word (in-list (cdr words))])
+        (display " ")
+        (display word))
+      (newline))
+    ; for-each is like map, but always returns (void)
+    (for-each display-words all-sorted-shifts))
+
+  (module+ test
+    (parameterize ([current-output-port (open-output-string)])
+      (kwik-display '(("A") ("B" "C")))
+      (check-equal?
+        (get-output-string (current-output-port))
+        "A\nB C\n")))
+
+  (: all-circular-shifts* : Lines -> (Listof Lines))
+  (define (all-circular-shifts* lines)
+    (map all-circular-shifts lines))
+
+  (module+ test
+    (check-equal?
+      (all-circular-shifts* '(("A" "B" "C") ("D")))
+      '((("C" "A" "B") ("B" "C" "A") ("A" "B" "C")) (("D")))))
+
+  (: kwik-index : Path-String -> Void)
+  (define (kwik-index file-name)
+    (define all-lines (kwik-split (kwik-read file-name)))
+    (define all-shifts (append* (all-circular-shifts* all-lines)))
+    (kwik-display (alphabetize all-shifts)))
+
+  (module+ test
+    (parameterize ([current-output-port (open-output-string)])
+      (define tmpfile (make-temporary-file))
+      (with-output-to-file tmpfile #:exists 'replace
+        (λ ()
+          (displayln "imagine if this")
+          (displayln "took 2 weeks to write")))
+      (kwik-index tmpfile)
+      (delete-file tmpfile)
+      (check-equal?
+        (get-output-string (current-output-port))
+        (string-join '(
+          "2 weeks to write took"
+          "if this imagine"
+          "imagine if this"
+          "this imagine if"
+          "to write took 2 weeks"
+          "took 2 weeks to write"
+          "weeks to write took 2"
+          "write took 2 weeks to\n") "\n"))))
+
+  (module+ main
+    (require racket/cmdline)
+    (: *output-to* (Parameterof Any))
+    (define *output-to* (make-parameter #f))
+    (command-line
+      #:program "kwik index"
+      #:once-each
+      [("-o" "--output")
+       output-to
+       "Write output to file"
+       (*output-to* output-to)]
+      #:args (file-name)
+      (define output-to (*output-to*))
+      (define out-port
+        (if (string? output-to)
+          (open-output-file output-to #:exists 'replace)
+          (current-output-port)))
+      (parameterize ([current-output-port out-port])
+        (kwik-index (cast file-name Path-String)))
+      (when (string? output-to)
+        (close-output-port out-port))))
+}
+
+Sample interactions:
+@codeblock{
+> racket kwik.rkt
+kwik index: expects 1 <file-name> on the command line, given 0 arguments
+
+> echo "It is a truth universally acknowledged" > pride-and-prejudice.txt
+> racket kwik.rkt -o index.out pride-and-prejudice.txt
+> wc -l index.out
+6 index.out
+}
+
+
+@section{Closed Parens}
+
+We started with functions, wrote (and quarantined) unit tests,
+ reinforced our design with types, and added a command-line interface.
+Going forward we could add Scribble documentation and share our work as a package.
```

Comment:
It would actually be pretty cool if you had those too, but maybe it'd be too long at that point. I guess a sequel is in order?